clippy_lints/src/methods/filter_map.rs

```diff
@@ -50,6 +52,197 @@ fn is_option_filter_map(cx: &LateContext<'_>, filter_arg: &hir::Expr<'_>, map_ar
+                // `if let Variant   (v) =         enum { v } else { unreachable!() }`
+                //         ^^^^^^^    ^            ^^^^            ^^^^^^^^^^^^^^^^^^
+                //    variant_span  variant_ident  scrutinee       else_ (blocks peeled later)
+                // OR
+                // `match enum {   Variant       (v) => v,      _ => unreachable!() }`
+                //        ^^^^     ^^^^^^^        ^                  ^^^^^^^^^^^^^^
+                //     scrutinee  variant_span  variant_ident        else_
+                let (scrutinee, else_, variant_ident, variant_span) =
+                    match higher::IfLetOrMatch::parse(cx, map_body.value) {
+                    Some(higher::IfLetOrMatch::IfLet(sc, pat, then, Some(else_)))
+                        if let Some((ident, span)) = expr_uses_local(pat, then) => {
+                        (sc, else_, ident, span)
+                    },
+                    Some(higher::IfLetOrMatch::Match(sc, [arm1, arm2], MatchSource::Normal)) => {
+                        let is_variant_arm = |arm: &hir::Arm<'_>, other_arm: &hir::Arm<'_>| {
+                            // check if the other arm has pattern `_` and this arm references the local (modulo blocks)
+                            if let PatKind::Wild = other_arm.pat.kind {
+                                expr_uses_local(arm.pat, arm.body.peel_blocks())
+                            } else {
+                                None
+                            }
+                        };
+                        let (else_arm, (ident, span)) = is_variant_arm(arm1, arm2)
+                            .map(|data| (arm2, data))
+                            .or_else(|| is_variant_arm(arm2, arm1).map(|data| (arm1, data)))?;
+
+                        (sc, else_arm.body, ident, span)
+                    },
+                    _ => return None,
+                };
+
+                if path_to_local_id(scrutinee, map_param_id)
+                    // else branch should be a `panic!` or `unreachable!` macro call
+                    && let Some(mac) = root_macro_call(else_.peel_blocks().span)
+                    && (is_panic(cx, mac.def_id) || cx.tcx.opt_item_name(mac.def_id) == Some(sym::unreachable))
+                {
+                    Some(CheckResult::PatternMatching { variant_span, variant_ident })
+                } else {
+                    None
+                }
+            },
+        }
+    }
+
+    fn hir(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, filter_param_id: HirId) -> Option<Self> {
+        if let ExprKind::MethodCall(path, receiver, [], _) = expr.kind
+            && let Some(recv_ty) = cx.typeck_results().expr_ty(receiver).peel_refs().ty_adt_def()
+            // TODO: we intentionally (?) do not check if `receiver` is a path expression, but maybe we should..?
+            // .filter(|x| effect(x).is_some()).map(|x| effect(x).unwrap()) to filter_map can have different behavior
```

Comment:
True, we could go with checking [`.can_have_side_effects()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Expr.html#method.can_have_side_effects) and if so mention it could change the behaviour in a note and adjust the applicability to `MaybeIncorrect`